ops/snapshot_mainchain.go

```diff
@@ -0,0 +1,218 @@
+	res, err := json.MarshalIndent(snapshot, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	file := fmt.Sprintf("%d-snapshot.json", latestBlock)
+	err = ioutil.WriteFile(file, res, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func snapshotInitializeCandidate(dposContract *mainchain.DPoS, start, end uint64) error {
+	log.Infof("Snapshot InitializeCandidate event from %d to %d", start, end)
+	initializeCandidateIt, err := dposContract.FilterInitializeCandidate(&bind.FilterOpts{
+		Start: start,
+		End:   &end,
+	}, []mainchain.Addr{})
+	if err != nil {
+		return err
+	}
+
+	defer initializeCandidateIt.Close()
+	for initializeCandidateIt.Next() {
+		initializeCandidate := initializeCandidateIt.Event
+		candidate := &Candidate{
+			MinSelfStake:    initializeCandidate.MinSelfStake,
+			CommissionRate:  initializeCandidate.CommissionRate,
+			RateLockEndTime: initializeCandidate.RateLockEndTime,
+			StakingPool:     big.NewInt(0),
+			Delegators:      make(map[string]*Delegator),
+		}
+		candidateMap[initializeCandidate.Candidate.String()] = candidate
+	}
+
+	return nil
+}
+
+func snapshotUpdateDelegatedStake(dposContract *mainchain.DPoS, start, end uint64) error {
+	log.Infof("Snapshot UpdateDelegatedStake event from %d to %d", start, end)
+	updateDelegatedStakeIt, err := dposContract.FilterUpdateDelegatedStake(&bind.FilterOpts{
+		Start: start,
+		End:   &end,
+	}, []mainchain.Addr{}, []mainchain.Addr{})
+	if err != nil {
+		return err
+	}
+
+	defer updateDelegatedStakeIt.Close()
+	for updateDelegatedStakeIt.Next() {
+		updateDelegatedStake := updateDelegatedStakeIt.Event
+		candidateAddr := updateDelegatedStake.Candidate.String()
+		delegatorAddr := updateDelegatedStake.Delegator.String()
+		candidate := candidateMap[candidateAddr]
+		_, ok := candidate.Delegators[delegatorAddr]
+		if !ok {
+			candidate.Delegators[delegatorAddr] = &Delegator{
+				DelegatedStake:    big.NewInt(0),
+				UndelegatingStake: big.NewInt(0),
+			}
+		}
+
+		_, ok = candidateDelegatorChanges[candidateAddr]
+		if !ok {
+			candidateDelegatorChanges[candidateAddr] = make(map[string]bool)
+		}
+
+		candidateDelegatorChanges[candidateAddr][delegatorAddr] = true
+	}
+
+	return nil
+}
+
+func syncCandidateDelegators(dposContract *mainchain.DPoS, blkNum *big.Int) error {
+	for candidateAddr, delegatorChanges := range candidateDelegatorChanges {
+		log.Infof("Snapshot candidate %s at %s", candidateAddr, blkNum)
+		candidateInfo, err := dposContract.GetCandidateInfo(&bind.CallOpts{BlockNumber: blkNum}, mainchain.Hex2Addr(candidateAddr))
+		if err != nil {
+			return err
+		}
+
+		candidate := candidateMap[candidateAddr]
+		candidate.CommissionRate = candidateInfo.CommissionRate
+		candidate.RateLockEndTime = candidateInfo.RateLockEndTime
+		candidate.StakingPool = candidateInfo.StakingPool
+		candidate.MinSelfStake = candidateInfo.MinSelfStake
+		candidate.Status = candidateInfo.Status
+		candidate.UnbondTime = candidateInfo.UnbondTime
```

Comment:
Values above (expect `StakingPool`) could be updated but not captured in `candidateDelegatorChanges` which is derived only from the `UpdateDelegatedStake` event. So we need either listen to those events update or scan the candidate list. Since the # of candidates is much smaller than # of delegators, we could just scan them. One way to do this is to initialize all candidates as entries in `candidateDelegatorChanges` after `snapshotInitializeCandidate`